### ISSUE TITLE
Eval nonfunctional?

### ISSUE THREAD

**Issue by ElfEars** [nesbox/TIC-80#800]
When attempting to eval code on the command line, nothing after the command is actually run.
I've attempted to modify the map and the sprite data of the loaded program with it and nothing happens.

I think there should be better cohesion between the command line and the code editor ala pico-8 (making it easier to perform poke/memcopy operations loaded programs) but that's probably for another issue (awkward due to the multiple languages. Maybe a command for each language ie: lua mset(0,0,32) ??)

**Comment by RobLoach** [comment]
eval code? What are you running?

**Comment by nesbox** [comment]
@RobLoach we have `eval` command in the console, you can pause a game and call code with `eval <code>` and `resume`, for example.
Implemented only in Lua atm.

**Comment by joshgoebel** [comment]
@ElfEars Did you `run` the cartridge first?  I found that some state seems to be necessary before eval does anything useful.

**Comment by atesin** [comment]
> @ElfEars Did you `run` the cartridge first? I found that some state seems to be necessary before eval does anything useful.

this should be the "accepted answer" ...  i had the same problem that eval "does nothing", until i `run` default cart then exit with [esc] key

until not the ideal condition... but to prevent head smashings in the while, it should be documented here anyway -> https://github.com/nesbox/TIC-80/wiki/Console#evaluating-code

(or better being fixed)

**Comment by technomancy** [comment]
I agree it's not very good the way it currently works without giving any feedback.

I tried to fix it, but unfortunately it just segfaults:

```
diff --git a/src/api/lua.c b/src/api/lua.c
index 696d8569..a9a633af 100644
--- a/src/api/lua.c
+++ b/src/api/lua.c
@@ -1778,7 +1778,10 @@ static void evalLua(tic_mem* tic, const char* code) {
     tic_core* core = (tic_core*)tic;
     lua_State* lua = core->currentVM;
 
-    if (!lua) return;
+    if (!lua) {
+      tic_api_trace(tic, "must run before eval", 1);
+      return;
+    }
 
     lua_settop(lua, 0);
 ```

I also tried this to automatically initialize, but it segfaults as well:

```
diff --git a/src/api/lua.c b/src/api/lua.c
index 696d8569..91a292cf 100644
--- a/src/api/lua.c
+++ b/src/api/lua.c
@@ -1778,7 +1778,10 @@ static void evalLua(tic_mem* tic, const char* code) {
     tic_core* core = (tic_core*)tic;
     lua_State* lua = core->currentVM;
 
-    if (!lua) return;
+    if (!lua) {
+      initLua(tic, "");
+      lua = core->currentVM = luaL_newstate();
+    }
 
     lua_settop(lua, 0);
```

Unfortunately I don't really know C, so I'm at a loss to explain what's going on. Maybe someone who knows C can help. =)

**Comment by icefoxen** [comment]
Thinking aloud...  `script_config->eval()` is called from `studio/screens/console.c` in `onEvalCommand()`, so if its segfaulting on `tic_api_trace()` then that means something in the `tic` object is not set up correctly before that is being called.  Looks like `tic_api_trace()` writes to `tic->data->trace()` which is initialized in `tic_core_tick()` which is called in `tic.c` and *deep breath* initialized to 

```c
static void onTrace(void* data, const char* text, u8 color)
{
    tic80* tic = (tic80*)data;

    if(tic->callback.trace)
        tic->callback.trace(text, color);
}
```

or else can be initialized in `studio/screens/run.c` and set to

```c
static void onTrace(void* data, const char* text, u8 color)
{
#if defined(BUILD_EDITORS)
    Run* run = (Run*)data;
    run->console->trace(run->console, text, color);
#endif
}
```

but that doesn't seem right 'cause I don't think a `Run*` is a valid `tic_mem*` object, so assuming it's the `tic_core_tick()` `onTrace()` impl  then `tic->callback` is set in `tic_core_tick()` to `config->callback` which pulls the correct trace callback from the `Languages[0]` array but not until *after* a cartridge is loaded otherwise `tic_core_tick()` just sits around doing nothing so the tldr is you can't call `tic_api_trace()` until a cartridge is loaded because it doesn't know which trace function to calllllllllllll

Also the bit in `core.c` line 506 that does `core->state.callback = config->callback` seems suspicious because `config` is a `tic_script_config` but the definition of that in `api.h` has a `callback` field that is a `tic_blit_callback` which doesn't seem to have a `trace` method, so it might be getting mixed up with the `trace` method on `callback` in the `tic80` structure defined in `include/tic80.h`.

I think.  This is probably all wrong.
